lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1556,6 +1557,28 @@ bool TypeChecker::coerceParameterListToType(ParameterList *P, DeclContext *DC,
     return hadError;
   };
 
+  if (!hadError && isa<ParenType>(paramListType.getPointer())) {
+    ParenType *parenTy = dyn_cast<ParenType>(paramListType.getPointer());
+    Type underlyingTy = parenTy->getUnderlyingType();
+    
+    while (isa<SubstitutedType>(underlyingTy.getPointer())) {
```

Comment:
You definitely want to be operating on a `CanType` rather than stripping away all of this yourself.